### PR TITLE
Upgrade to latest version of ConfigurationBinder class as of .net 8

### PR DESCRIPTION
**Pull request by helto4real**
## Proposed change

Upgraded the ConfigurationBinder to the .net 8 version.
Due to this:
- We now support integer type enums
- We now support read only collection types that have a init property and is initialized

## Type of change

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information

- This PR fixes or closes issue: fixes #
- This PR is related to issue:
- Link to documentation pull request:

## Checklist

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] The code compiles without warnings (code quality check)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration are added/changed:

- [ ] Documentation added/updated for http://netdaemon.xtz

[Docs repository]: https://github.com/net-daemon/docs
[Developer documentation]: https://netdaemon.xyz/docs/developer/

### CODE REVIEW

**Comment by helto4real** [comment]
The test coverage is due to using Microsoft code. I really do not want to test that again :). I am find with that.

**Comment by helto4real** [comment]
image
Strange, the reporting of check got bugged after coverage was reported. The build and tests are ok.

File: src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs [hunk 1]
```diff
@@ -334,36 +913,53 @@ private bool TryConvertValue(Type type, string value, out object? result, out Ex
...
         // No standard converter is available so lets´s try find
         // a registered service that converts to the type from string
 
         try
         {
-            result = ActivatorUtilities.CreateInstance(_provider, type, value);
+            result = ActivatorUtilities.CreateInstance(provider, type, value);
         }
         catch (Exception e)
         {
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs:965]: In the other cases above the new exception is assigned to `error` . Not sure why, but maybe its good to do the same here

File: src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs [hunk 2]
```diff
@@ -11,320 +20,890 @@ internal interface IConfigurationBinding
...
 /// <summary>
-///     Class that allows binding strongly typed objects to configuration values.
+/// Wrapper around the ConfigurationBinder to make it available from the service provider
+/// and to inject the IServiceProvider to allow instancing objects that are registered services
 /// </summary>
-/// <remarks>
-///     This is a scaled down and modified version of .NET 7 implementation.
-/// </remarks>
-internal class ConfigurationBinding : IConfigurationBinding
+internal class ConfigurationBinding(IServiceProvider provider) : IConfigurationBinding
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs:26]: Keep in a separete file, so we keep the copied stuff as original as possible. Maybe adding some marker comments like 
// ****** NetDaemon customisation ******
where we added things
And maybe add a link to the specific version of the original file and  you used to copy this from, so in the future we can diff to see what exactly we changed

File: src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs (hunk 1 above)
Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs]: Done, probably a good thing to do. :)

File: src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs (hunk 2 above)
Comment (helto4real) [reply src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinding.cs:9]: Yea split all interfaces and classes plus the link now

File: src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinder.cs [hunk 3]
```diff
@@ -0,0 +1,1118 @@
+// This is a modified version of the .net 8 source code for ConfigurationBinder.cs
+// https://github.com/dotnet/runtime/blob/v8.0.7/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.class
```
Comment (FrankBakkerNl) [inline src/AppModel/NetDaemon.AppModel/Internal/Config/ConfigurationBinder.cs:2]: Link doe snot seem to work

```suggestion
// https://github.com/dotnet/runtime/blob/2aade6beb02ea367fd97c4070a4198802fe61c03/src/libraries/Microsoft.Extensions.Configuration.Binder/src/ConfigurationBinder.cs
```

**Review by FrankBakkerNl — APPROVED** [review]

**Pull request merged**